### ISSUE TITLE
Edge case problem with NullBackend for cache_core

### ISSUE THREAD

**Issue by NamelessCoder** [helhum/TYPO3-Distribution#9]
In the file https://github.com/helhum/TYPO3-Distribution/blob/master/conf/dev.config.cache.yml#L4 the cache backend for cache_core is set to the NullBackend, but unfortunately this causes some unexpected problems with TCA loading.

I faced a use case in a current project where an Extbase BE module used three different entities, all configured the exact same way in TCA with proper file-convention-based inclusion (not overrides). Strangely enough, suddenly after an update of an ext *one* of those entities would fail to map properties. After digging for a while I found that this was because the TCA array for that table was completely missing, simply not being loaded. It was there for others, but not for that single table.

While I haven't tracked down the exact reason for this, I suggest changing this one particular cache to use the SimpleFileBackend as it does in production, since it can't use the TransientMemoryBackend I think - the cache pretty much assumes you use a *FileBackend that stores PHP code.

**Comment by helhum** [comment]
Hey Claus.

> the cache backend for cache_core is set to the NullBackend, but unfortunately this causes some unexpected problems with TCA loading.

Not having a cache should not affect the application from working. Slowing down, yes. Not working, nope.

> While I haven't tracked down the exact reason for this

That is what we desperately need to find out. Just using a cache (which is very annoying during development) is not feasible imho.

If you have anything to reproduce the faulty behavior (I have no clue what can go wrong here), go ahead.

**Comment by NamelessCoder** [comment]
> Not having a cache should not affect the application from working. Slowing down, yes. Not working, nope.

I completely agree with this expectation, but that is not what I found after two hours of debugging. There is literally nothing special about this extension's TCA, yet that single entity had no TCA in the BE module.

I don't know it if is reproducible with that extension alone, but the one that caused the symptom for me was: https://github.com/mia3/crossmedia_fourallportal (at dev-master).

The entity that doesn't have any TCA in the BE module is the "Server" - this can be added only on the root page and can be filled with garbage info for testing. The BE module that exhibits the problem is the "4Allportal Connector", when switching to the "Servers" listing. Here, each "Server" has no attributes because the entity has no properties filled, which is caused by the TCA table missing (this can be confirmed in the persistence backend: the record that gets pulled is filled correctly, but the properties get filtered out by the TCA check, causing them to not be persisted and not be thawed from persistence).

Caveat: there are other extensions installed of course, which I can't share.

I hope this helps. Unfortunately I'm not able to further track down the cause since this is a paid project and the problem is already solved by the cache thing (which is feasible to us in this project at least).

**Comment by NamelessCoder** [comment]
I found the problem, leaving the explanation here for anyone searching for a similar issue:

The TCA file called a function from another extension which analysed TCA from that extension, but because the first extension did not declare that it depends on the second one, the loading order was incorrect and so the TCA from the second extension was not available. Declaring the dependency properly resulted in a correct loading order which meant the TCA was available as expected.

The reason this worked differently with a generated TCA cache is most likely that the first pass wrote the TCA like it should but without the special things added by the second extension, then a second system cache flush re-wrote that TCA but did not unset the TCA global before doing so - finally resulting in a working state of the TCA despite the incorrect loading order of extensions' TCA files.

This is pretty obvious when you look at it now, but the symptom this error had was really bizarre. Sorry about the unrelated bug report but hopefully this saves someone else from the same digging trip I just went on!
